File: swift-bridge/Sources/CoreImageBridge/Vector.swift
import CoreGraphics
import CoreImage
import Foundation

@_cdecl("ci_vector_new1")
public func ci_vector_new1(_ x: Double) -> UnsafeMutableRawPointer? {
    ci_retain(CIVector(x: x))
}

@_cdecl("ci_vector_new2")
public func ci_vector_new2(_ x: Double, _ y: Double) -> UnsafeMutableRawPointer? {
    ci_retain(CIVector(x: x, y: y))
}

@_cdecl("ci_vector_new3")
public func ci_vector_new3(_ x: Double, _ y: Double, _ z: Double) -> UnsafeMutableRawPointer? {
    ci_retain(CIVector(x: x, y: y, z: z))
}

@_cdecl("ci_vector_new4")
public func ci_vector_new4(
    _ x: Double,
    _ y: Double,
    _ z: Double,
    _ w: Double
) -> UnsafeMutableRawPointer? {
    ci_retain(CIVector(x: x, y: y, z: z, w: w))
}

@_cdecl("ci_vector_from_rect")
public func ci_vector_from_rect(_ x: Double, _ y: Double, _ width: Double, _ height: Double) -> UnsafeMutableRawPointer? {
    ci_retain(CIVector(cgRect: CGRect(x: x, y: y, width: width, height: height)))
}

@_cdecl("ci_vector_from_transform")
public func ci_vector_from_transform(
    _ a: Double,
    _ b: Double,
    _ c: Double,
    _ d: Double,
    _ tx: Double,
    _ ty: Double
) -> UnsafeMutableRawPointer? {
    let transform = CGAffineTransform(a: a, b: b, c: c, d: d, tx: tx, ty: ty)
    return ci_retain(CIVector(cgAffineTransform: transform))
}

@_cdecl("ci_vector_from_string")
public func ci_vector_from_string(_ representation: UnsafePointer<CChar>?) -> UnsafeMutableRawPointer? {
    guard let representation else { return nil }
    return ci_retain(CIVector(string: String(cString: representation)))
}

@_cdecl("ci_vector_count")
public func ci_vector_count(_ handle: UnsafeMutableRawPointer?) -> Int {
    guard let vector: CIVector = ci_borrow(handle) else { return 0 }
    return Int(vector.count)
}

@_cdecl("ci_vector_value_at")
public func ci_vector_value_at(_ handle: UnsafeMutableRawPointer?, _ index: Int) -> Double {
    guard let vector: CIVector = ci_borrow(handle), index >= 0 else { return 0 }
    return Double(vector.value(at: index))
}

@_cdecl("ci_vector_x")
public func ci_vector_x(_ handle: UnsafeMutableRawPointer?) -> Double {
    guard let vector: CIVector = ci_borrow(handle) else { return 0 }
    return Double(vector.x)
}

@_cdecl("ci_vector_y")
public func ci_vector_y(_ handle: UnsafeMutableRawPointer?) -> Double {
    guard let vector: CIVector = ci_borrow(handle) else { return 0 }
    return Double(vector.y)
}

@_cdecl("ci_vector_z")
public func ci_vector_z(_ handle: UnsafeMutableRawPointer?) -> Double {
    guard let vector: CIVector = ci_borrow(handle) else { return 0 }
    return Double(vector.z)
}

@_cdecl("ci_vector_w")
public func ci_vector_w(_ handle: UnsafeMutableRawPointer?) -> Double {
    guard let vector: CIVector = ci_borrow(handle) else { return 0 }
    return Double(vector.w)
}

@_cdecl("ci_vector_point")
public func ci_vector_point(
    _ handle: UnsafeMutableRawPointer?,
    _ outX: UnsafeMutablePointer<Double>?,
    _ outY: UnsafeMutablePointer<Double>?
) {
    guard let vector: CIVector = ci_borrow(handle) else {
        outX?.pointee = 0
        outY?.pointee = 0
        return
    }
    let point = vector.cgPointValue
    outX?.pointee = point.x
    outY?.pointee = point.y
}

@_cdecl("ci_vector_rect")
public func ci_vector_rect(
    _ handle: UnsafeMutableRawPointer?,
    _ outX: UnsafeMutablePointer<Double>?,
    _ outY: UnsafeMutablePointer<Double>?,
    _ outWidth: UnsafeMutablePointer<Double>?,
    _ outHeight: UnsafeMutablePointer<Double>?
) {
    guard let vector: CIVector = ci_borrow(handle) else {
        outX?.pointee = 0
        outY?.pointee = 0
        outWidth?.pointee = 0
        outHeight?.pointee = 0
        return
    }
    let rect = vector.cgRectValue
    outX?.pointee = rect.origin.x
    outY?.pointee = rect.origin.y
    outWidth?.pointee = rect.size.width
    outHeight?.pointee = rect.size.height
}

@_cdecl("ci_vector_transform")
public func ci_vector_transform(
    _ handle: UnsafeMutableRawPointer?,
    _ outA: UnsafeMutablePointer<Double>?,
    _ outB: UnsafeMutablePointer<Double>?,
    _ outC: UnsafeMutablePointer<Double>?,
    _ outD: UnsafeMutablePointer<Double>?,
    _ outTx: UnsafeMutablePointer<Double>?,
    _ outTy: UnsafeMutablePointer<Double>?
) {
    guard let vector: CIVector = ci_borrow(handle) else {
        outA?.pointee = 0
        outB?.pointee = 0
        outC?.pointee = 0
        outD?.pointee = 0
        outTx?.pointee = 0
        outTy?.pointee = 0
        return
    }
    let transform = vector.cgAffineTransformValue
    outA?.pointee = transform.a
    outB?.pointee = transform.b
    outC?.pointee = transform.c
    outD?.pointee = transform.d
    outTx?.pointee = transform.tx
    outTy?.pointee = transform.ty
}

@_cdecl("ci_vector_string_representation")
public func ci_vector_string_representation(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let vector: CIVector = ci_borrow(handle) else { return nil }
    return ci_string(vector.stringRepresentation)
}

File: src/vector.rs
use core::ffi::c_void;
use core::fmt;
use core::ptr;

use apple_cf::cg::{CGAffineTransform, CGPoint, CGRect};

use crate::ffi;
use crate::util::{string_to_cstring, take_owned_string};

/// A Core Image vector value used for filter inputs such as centers, extents, and corners.
pub struct CIVector {
    ptr: *mut c_void,
}

impl Drop for CIVector {
    fn drop(&mut self) {
        if !self.ptr.is_null() {
            unsafe { ffi::ci_object_release(self.ptr) };
            self.ptr = ptr::null_mut();
        }
    }
}

impl Clone for CIVector {
    fn clone(&self) -> Self {
        Self {
            ptr: unsafe { ffi::ci_object_retain(self.ptr) },
        }
    }
}

impl fmt::Debug for CIVector {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("CIVector")
            .field("ptr", &self.ptr)
            .field("values", &self.values())
            .finish_non_exhaustive()
    }
}

impl CIVector {
    pub(crate) const unsafe fn from_raw(ptr: *mut c_void) -> Self {
        Self { ptr }
    }

    fn from_non_null(ptr: *mut c_void, kind: &str) -> Self {
        assert!(!ptr.is_null(), "{kind} returned nil");
        unsafe { Self::from_raw(ptr) }
    }

    pub const fn as_ptr(&self) -> *mut c_void {
        self.ptr
    }

    pub fn new1(x: f64) -> Self {
        Self::from_non_null(unsafe { ffi::ci_vector_new1(x) }, "CIVector")
    }

    pub fn new(x: f64, y: f64) -> Self {
        Self::from_non_null(unsafe { ffi::ci_vector_new2(x, y) }, "CIVector")
    }

    pub fn new3(x: f64, y: f64, z: f64) -> Self {
        Self::from_non_null(unsafe { ffi::ci_vector_new3(x, y, z) }, "CIVector")
    }

    pub fn new4(x: f64, y: f64, z: f64, w: f64) -> Self {
        Self::from_non_null(unsafe { ffi::ci_vector_new4(x, y, z, w) }, "CIVector")
    }

    pub fn from_point(point: CGPoint) -> Self {
        Self::new(point.x, point.y)
    }

    pub fn from_rect(rect: CGRect) -> Self {
        Self::from_non_null(
            unsafe { ffi::ci_vector_from_rect(rect.origin.x, rect.origin.y, rect.size.width, rect.size.height) },
            "CIVector",
        )
    }

    pub fn from_transform(transform: CGAffineTransform) -> Self {
        Self::from_non_null(
            unsafe {
                ffi::ci_vector_from_transform(
                    transform.a,
                    transform.b,
                    transform.c,
                    transform.d,
                    transform.tx,
                    transform.ty,
                )
            },
            "CIVector",
        )
    }

    pub fn from_string(representation: &str) -> Option<Self> {
        let representation = string_to_cstring(representation, "representation").ok()?;
        let handle = unsafe { ffi::ci_vector_from_string(representation.as_ptr()) };
        if handle.is_null() {
            None
        } else {
            Some(unsafe { Self::from_raw(handle) })
        }
    }

    pub fn count(&self) -> usize {
        unsafe { ffi::ci_vector_count(self.ptr) }
    }

    pub fn value_at(&self, index: usize) -> f64 {
        unsafe { ffi::ci_vector_value_at(self.ptr, index) }
    }

    pub fn values(&self) -> Vec<f64> {
        (0..self.count())
            .map(|index| self.value_at(index))
            .collect()
    }

    pub fn x(&self) -> f64 {
        unsafe { ffi::ci_vector_x(self.ptr) }
    }

    pub fn y(&self) -> f64 {
        unsafe { ffi::ci_vector_y(self.ptr) }
    }

    pub fn z(&self) -> f64 {
        unsafe { ffi::ci_vector_z(self.ptr) }
    }

    pub fn w(&self) -> f64 {
        unsafe { ffi::ci_vector_w(self.ptr) }
    }

    pub fn point_value(&self) -> CGPoint {
        let mut x = 0.0;
        let mut y = 0.0;
        unsafe { ffi::ci_vector_point(self.ptr, &mut x, &mut y) };
        CGPoint { x, y }
    }

    pub fn rect_value(&self) -> CGRect {
        let mut x = 0.0;
        let mut y = 0.0;
        let mut width = 0.0;
        let mut height = 0.0;
        unsafe { ffi::ci_vector_rect(self.ptr, &mut x, &mut y, &mut width, &mut height) };
        CGRect::new(x, y, width, height)
    }

    pub fn affine_transform_value(&self) -> CGAffineTransform {
        let mut a = 0.0;
        let mut b = 0.0;
        let mut c = 0.0;
        let mut d = 0.0;
        let mut tx = 0.0;
        let mut ty = 0.0;
        unsafe {
            ffi::ci_vector_transform(self.ptr, &mut a, &mut b, &mut c, &mut d, &mut tx, &mut ty);
        };
        CGAffineTransform { a, b, c, d, tx, ty }
    }

    pub fn string_representation(&self) -> String {
        unsafe { take_owned_string(ffi::ci_vector_string_representation(self.ptr)) }
            .unwrap_or_default()
    }
}

impl From<(f64,)> for CIVector {
    fn from(value: (f64,)) -> Self {
        Self::new1(value.0)
    }
}

impl From<(f64, f64)> for CIVector {
    fn from(value: (f64, f64)) -> Self {
        Self::new(value.0, value.1)
    }
}

impl From<(f64, f64, f64)> for CIVector {
    fn from(value: (f64, f64, f64)) -> Self {
        Self::new3(value.0, value.1, value.2)
    }
}

impl From<(f64, f64, f64, f64)> for CIVector {
    fn from(value: (f64, f64, f64, f64)) -> Self {
        Self::new4(value.0, value.1, value.2, value.3)
    }
}
